int64 id
float64 p
int64 x
int64 y